### PR TITLE
2D canvas and layout (WIP)

### PR DESCRIPTION
**Pull request by AmirArdroudi**

### CODE REVIEW

File: src/2d/CMakeLists.txt [hunk 1]
```diff
@@ -2,12 +2,15 @@ fips_begin_module(ari_2d)
...
+        Sprite.hpp				  Sprite.cpp
 		Camera2D.hpp              Camera2D.cpp		
-        FrameData2D.hpp			 
         Node2D.hpp
 		RenderSystem2D.hpp        RenderSystem2D.cpp
         SceneSystem2D.hpp         SceneSystem2D.cpp
-        Sprite.hpp                Sprite.cpp
+		CanvasNode.cpp			  CanvasNode.hpp
+		Layout.cpp				  Layout.hpp
+
```
Comment (kochol) [inline src/2d/CMakeLists.txt:13]: Sort the files by name.

File: src/gfx/gfx.cpp [hunk 2]
```diff
@@ -533,13 +536,55 @@ namespace ari
...
+		sx_rect GetScissorRect()
+		{
+			return g_rectScissor;
+		}
+		//------------------------------------------------------------------------------
+		void SetViewport(Viewport* p_viewport, bool origin_top_left)
+		{
+			sg_apply_viewport(p_viewport->Rect.x, p_viewport->Rect.y, p_viewport->Rect.width, p_viewport->Rect.height, origin_top_left);
+			      g_Viewport = p_viewport;
+		}
```
Comment (kochol) [inline src/gfx/gfx.cpp:565]: You forget to set g_rectVP

File: src/gfx/gfx.cpp [hunk 3]
```diff
@@ -98,6 +98,9 @@ namespace ari
 		core::Map<uint32_t, ShaderDescShaderHandle> MaterialShaders;
 
 		static sx_mat4 g_mView, g_mProj, g_mViewProj, g_mWorld, g_mWorldViewProj;
+		static Viewport* g_Viewport = nullptr;
```
Comment (kochol) [inline src/gfx/gfx.cpp:101]: Use pointer is not safe here

File: src/gfx/gfx.cpp [hunk 4]
```diff
@@ -533,13 +536,55 @@ namespace ari
...
+			g_rectScissor.xmax = width;
+			g_rectScissor.ymax = height;
+		}
+
+		sx_rect GetScissorRect()
+		{
+			return g_rectScissor;
+		}
+		//------------------------------------------------------------------------------
+		void SetViewport(Viewport* p_viewport, bool origin_top_left)
```
Comment (kochol) [inline src/gfx/gfx.cpp:561]: Rename to SetViewportRect and GetViewportRect

File: src/2d/CMakeLists.txt (hunk 1 above)
Comment (AmirArdroudi) [reply src/2d/CMakeLists.txt:13]: done. When or where should I use sx_rect instead of air::core::RectI ?

Comment (AmirArdroudi) [reply src/2d/CMakeLists.txt:13]: done! and where or when should I use ari::core::RectI instead of sx_rect in general?

Comment (kochol) [reply src/2d/CMakeLists.txt:13]: Always

Comment (kochol) [reply src/2d/CMakeLists.txt:13]: I changed my mind, delete RectI completely and use sx_rect every where.

Comment (AmirArdroudi) [reply src/2d/CMakeLists.txt:13]: ok sure

File: src/gfx/gfx.cpp (hunk 2 above)
Comment (AmirArdroudi) [reply src/gfx/gfx.cpp:565]: fixed

File: src/gfx/gfx.cpp (hunk 3 above)
Comment (AmirArdroudi) [reply src/gfx/gfx.cpp:101]: why not? what if I set it as a const pointer?

File: src/2d/CMakeLists.txt (hunk 1 above)
Comment (AmirArdroudi) [reply src/2d/CMakeLists.txt:13]: Then should I change Rect in viewport.hpp to sx_rect type?

Comment (kochol) [reply src/2d/CMakeLists.txt:13]: yes

File: src/gfx/gfx.cpp (hunk 4 above)
Comment (AmirArdroudi) [reply src/gfx/gfx.cpp]: fixed

File: src/2d/CMakeLists.txt (hunk 1 above)
Comment (AmirArdroudi) [reply src/2d/CMakeLists.txt]: applied!

File: src/2d/RenderSystem2D.cpp [hunk 5]
```diff
@@ -39,6 +39,9 @@ namespace ari::en
 				// TODO: Set viewport
 				gfx::Viewport* pViewport = m_pFrameDataCurrent->Camera2dObj->GetViewport();
 
+				sx_rect ViewportRect = m_pFrameDataCurrent->Camera2dObj->GetViewportRect();
+				gfx::SetViewportRect(ViewportRect, false);
+
```
Comment (kochol) [inline src/2d/RenderSystem2D.cpp:44]: What if the user didn't set the viewport rect at all.

File: src/2d/SceneSystem2D.cpp [hunk 6]
```diff
@@ -60,7 +60,7 @@ namespace ari::en
 					rect = io::GetWindowSize(TargetWindow);
 				}
 
-				m_pActiveCamera2D->_proj = sx_mat4_ortho(rect.width, rect.height, 0.1f, 100, 0, false);
+				m_pActiveCamera2D->_proj = sx_mat4_ortho(rect.xmax, rect.ymax, 0.1f, 100, 0, false);
```
Comment (kochol) [inline src/2d/SceneSystem2D.cpp:63]: set znear to 0.0f

File: src/2d/RenderSystem2D.cpp (hunk 5 above)
Comment (AmirArdroudi) [reply src/2d/RenderSystem2D.cpp:44]: then we should set window rect as default, yeah?

Comment (kochol) [reply src/2d/RenderSystem2D.cpp:44]: yes

**Pull request closed without merging**